File: stack/python/befitting_brackets.py
from typing import Iterable


def befitting_brackets(s: str) -> bool:
    stack: list[str] = []
    matching = {')': '(', ']': '[', '}': '{'}

    for ch in s:
        if ch in '([{':
            stack.append(ch)
        elif ch in matching:
            if not stack or stack.pop() != matching[ch]:
                return False
    return not stack

File: stack/csharp/BefittingBrackets.cs
namespace StackSolutions;

public static class BefittingBrackets
{
    public static bool IsValid(string s)
    {
        var stack = new Stack<char>();
        foreach (var ch in s)
        {
            switch (ch)
            {
                case '(':
                case '[':
                case '{':
                    stack.Push(ch);
                    break;
                case ')':
                case ']':
                case '}':
                    if (stack.Count == 0 || stack.Pop() != Matching(ch))
                    {
                        return false;
                    }
                    break;
            }
        }

        return stack.Count == 0;
    }

    private static char Matching(char closing) => closing switch
    {
        ')' => '(',
        ']' => '[',
        '}' => '{',
        _ => '?'
    };
}
